File: services/api-gateway/routes/digest.py
"""
Digest routes for the API Gateway - proxies to Data Server
"""

import logging
from typing import Dict, Any, Optional, List

import httpx
from fastapi import APIRouter, HTTPException, Depends, Query, status
from pydantic import BaseModel

from auth import get_current_user
from config import settings
from auth import create_internal_api_headers

logger = logging.getLogger(__name__)
router = APIRouter()


# Pydantic models
class DigestResponse(BaseModel):
    """Response model for digest data"""
    success: bool
    data: Dict[str, Any]


class DigestListResponse(BaseModel):
    """Response model for digest lists"""
    success: bool
    data: Dict[str, Any]
    total: Optional[int] = None


class EmailDigestRequest(BaseModel):
    """Request model for creating email digests"""
    emails: List[Dict[str, Any]]


class ThematicProcessRequest(BaseModel):
    """Request model for thematic processing"""
    emails: List[Dict[str, Any]]
    email_digest_id: Optional[int] = None


# Helper function to proxy requests to Data Server
async def proxy_to_data_server(
    method: str,
    path: str,
    json_data: Optional[Dict] = None,
    params: Optional[Dict] = None
) -> Dict[str, Any]:
    """Proxy request to Data Server with internal authentication"""
    
    url = f"{settings.DATA_SERVER_URL.rstrip('/')}/api/{path.lstrip('/')}"
    headers = create_internal_api_headers()
    
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.request(
                method=method,
                url=url,
                json=json_data,
                params=params,
                headers=headers
            )
            
            if response.status_code >= 400:
                logger.error(f"Data Server error: {response.status_code} - {response.text}")
                raise HTTPException(
                    status_code=response.status_code,
                    detail=response.json().get("message", "Service error")
                )
            
            return response.json()
            
    except httpx.RequestError as e:
        logger.error(f"Request error to Data Server: {e}")
        raise HTTPException(
            status_code=503,
            detail=f"Data Server unavailable: {str(e)}"
        )


@router.get("/latest", response_model=DigestResponse)
async def get_latest_digest(
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Get the latest digest for the current user
    
    Prioritizes thematic digests, falls back to regular digests
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "GET",
            f"digest/latest/{user_id}"
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting latest digest: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve latest digest"
        )


@router.get("/detailed", response_model=DigestResponse)
async def get_detailed_digest(
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Get the latest detailed digest (individual emails) for the current user
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "GET",
            f"digest/detailed/{user_id}"
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting detailed digest: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve detailed digest"
        )


@router.get("/history", response_model=DigestListResponse)
async def get_digest_history(
    current_user: Dict[str, Any] = Depends(get_current_user),
    limit: Optional[int] = Query(None, description="Number of digests to return"),
    offset: Optional[int] = Query(None, description="Number of digests to skip")
):
    """
    Get digest history for the current user
    """
    user_id = current_user["uid"]
    
    try:
        params = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        
        result = await proxy_to_data_server(
            "GET",
            f"digest/history/{user_id}",
            params=params
        )
        
        return DigestListResponse(
            success=True,
            data=result.get("data", {}),
            total=result.get("data", {}).get("total")
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting digest history: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve digest history"
        )


@router.get("/date/{date}", response_model=DigestResponse)
async def get_digest_by_date(
    date: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Get digest for a specific date (YYYY-MM-DD format)
    
    Returns thematic digest if available, otherwise regular digest
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "GET",
            f"digest/date/{user_id}/{date}"
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting digest by date: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve digest for date"
        )


@router.post("/create", response_model=DigestResponse)
async def create_digest(
    request: EmailDigestRequest,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Create a new digest from provided emails
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "POST",
            "digest/create",
            json_data={
                "user_id": user_id,
                "emails": request.emails
            }
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating digest: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create digest"
        )


@router.post("/generate", response_model=DigestResponse)
async def generate_digest(
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Generate a new digest automatically by fetching recent emails
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "POST",
            "digest/generate",
            json_data={
                "user_id": user_id
            }
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error generating digest: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to generate digest"
        )


@router.post("/thematic/process", response_model=DigestResponse)
async def process_thematic_digest(
    request: ThematicProcessRequest,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Process emails into a thematic digest using AI analysis
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "POST",
            "digest/thematic/process",
            json_data={
                "userId": user_id,
                "emails": request.emails,
                "emailDigestId": request.email_digest_id
            }
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error processing thematic digest: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to process thematic digest"
        )


@router.get("/stats", response_model=DigestResponse)
async def get_digest_stats(
    current_user: Dict[str, Any] = Depends(get_current_user),
    period: Optional[str] = Query(None, description="Time period: day, week, month, year")
):
    """
    Get digest statistics for the current user
    """
    user_id = current_user["uid"]
    
    try:
        params = {}
        if period:
            params["period"] = period
        
        result = await proxy_to_data_server(
            "GET",
            f"digest/stats/{user_id}",
            params=params
        )
        
        return DigestResponse(
            success=True,
            data=result.get("data", {})
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting digest stats: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve digest statistics"
        )


@router.get("/available-dates", response_model=DigestResponse)
async def get_available_digest_dates(
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Get all dates that have digests available for the current user
    """
    user_id = current_user["uid"]
    
    try:
        result = await proxy_to_data_server(
            "GET",
            f"storage/available-digest-dates/{user_id}"
        )
        
        return DigestResponse(
            success=True,
            data={"dates": result.get("data", [])}
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting available digest dates: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve available digest dates"
        )

File: services/data-server/src/routes/digest.ts
/**
 * Digest Routes - AI Processing and Business Logic
 * 
 * Handles digest generation, OpenAI processing, and thematic analysis
 */

import { Router, Request, Response } from 'express';
import { asyncHandler } from '../middleware/error';
import { generateDigest, getLatestDigest, getLatestThematicDigest } from '../services/openai';
import { storage } from '../services/storage';
import { queueDigestGeneration } from '../services/celery-client';

const router = Router();

// Helper function for standard API responses
const apiResponse = (data: any, message?: string) => ({
  success: true,
  data,
  ...(message && { message })
});

const apiError = (message: string, code?: string) => ({
  success: false,
  error: message,
  ...(code && { code })
});

// ==================== DIGEST GENERATION ====================

// Create digest from processed emails
router.post('/create', asyncHandler(async (req: Request, res: Response) => {
  const { user_id, emails } = req.body;
  
  if (!user_id || !emails || !Array.isArray(emails)) {
    return res.status(400).json(apiError('user_id and emails array are required', 'MISSING_FIELDS'));
  }
  
  console.log(`📧 Creating digest for user ${user_id} with ${emails.length} emails`);
  
  try {
    // Transform incoming email data to EmailInput format
    const transformedEmails = emails.map((email: any) => ({
      sender: email.sender,
      subject: email.subject,
      content: email.content,
      receivedAt: new Date(email.received_at || email.receivedAt), // Handle both snake_case and camelCase
      originalLink: email.original_link || email.originalLink
    }));
    
    // Generate digest using existing OpenAI service
    const digest = await generateDigest(user_id, transformedEmails);
    
    res.status(201).json(apiResponse(digest, 'Digest created successfully'));
    
  } catch (error: any) {
    console.error('Error creating digest:', error);
    res.status(500).json(apiError(
      `Failed to create digest: ${error.message}`,
      'DIGEST_CREATION_FAILED'
    ));
  }
}));

// Generate digest automatically (new simple endpoint)
router.post('/generate', asyncHandler(async (req: Request, res: Response) => {
  const { user_id } = req.body;

  if (!user_id) {
    return res.status(400).json(apiError('user_id is required', 'MISSING_FIELDS'));
  }

  console.log(`🚀 Auto-generating digest for user ${user_id}`);

  try {
    // Get monitored emails for the user
    const monitoredEmails = await storage.getMonitoredEmails(user_id);

    if (monitoredEmails.length === 0) {
      return res.status(400).json(apiError('No monitored emails configured. Please add email sources first.', 'NO_MONITORED_EMAILS'));
    }

    // Get OAuth tokens for Gmail access
    const oauthTokens = await storage.getOAuthTokenByUid(user_id);

    if (!oauthTokens) {
      return res.status(400).json(apiError('No OAuth tokens found. Please re-authenticate with Gmail.', 'NO_OAUTH_TOKENS'));
    }

    // Queue the digest generation task to Celery worker
    const taskId = await queueDigestGeneration(user_id);

    res.json(apiResponse({
      message: 'Digest generation started successfully',
      user_id,
      monitored_emails_count: monitoredEmails.length,
      status: 'processing',
      task_id: taskId
    }, 'Digest generation initiated'));

  } catch (error: any) {
    console.error('Error auto-generating digest:', error);
    res.status(500).json(apiError(
      `Failed to generate digest: ${error.message}`,
      'DIGEST_GENERATION_FAILED'
    ));
  }
}));

// Generate digest manually (for API gateway)
router.post('/generate/:userId', asyncHandler(async (req: Request, res: Response) => {
  const { userId } = req.params;
  const { monitored_emails, oauth_data } = req.body;
  
  if (!monitored_emails || !Array.isArray(monitored_emails)) {
    return res.status(400).json(apiError('monitored_emails array is required', 'MISSING_FIELDS'));
  }
  
  console.log(`🔄 Manual digest generation for user ${userId}`);
  
  try {
    // This would typically trigger the email worker
    // For now, return a placeholder response
    res.json(apiResponse({
      message: 'Digest generation triggered',
      userId,
      monitoredEmails: monitored_emails.length,
      status: 'processing'
    }, 'Digest generation started'));
    
  } catch (error: any) {
    console.error('Error triggering digest generation:', error);
    res.status(500).json(apiError(
      `Failed to trigger digest generation: ${error.message}`,
      'DIGEST_GENERATION_FAILED'
    ));
  }
}));

// ==================== DIGEST RETRIEVAL ====================

// Get latest digest (prioritizes thematic)
router.get('/latest/:userId', asyncHandler(async (req: Request, res: Response) => {
  const { userId } = req.params;
  
  console.log(`📊 Getting latest digest for user ${userId}`);
  
  try {
    // Try thematic digest first, fall back to regular
    const latestDigest = await getLatestThematicDigest(userId);
    
    if (!latestDigest) {
      return res.status(404).json(apiError('No digests found for user', 'NOT_FOUND'));
    }
    
    res.json(apiResponse(latestDigest));
    
  } catch (error: any) {
    console.error('Error getting latest digest:', error);
    res.status(500).json(apiError(
      `Failed to get latest digest: ${error.message}`,
      'DIGEST_RETRIEVAL_FAILED'
    ));
  }
}));

// Get latest detailed digest (individual emails)
router.get('/detailed/:userId', asyncHandler(async (req: Request, res: Response) => {
  const { userId } = req.params;
  
  console.log(`📧 Getting detailed digest for user ${userId}`);
  
  try {
    const detailedDigest = await getLatestDigest(userId);
    
    if (!detailedDigest) {
      return res.status(404).json(apiError('No detailed digests found for user', 'NOT_FOUND'));
    }
    
    res.json(apiResponse(detailedDigest));
    
  } catch (error: any) {
    console.error('Error getting detailed digest:', error);
    res.status(500).json(apiError(
      `Failed to get detailed digest: ${error.message}`,
      'DIGEST_RETRIEVAL_FAILED'
    ));
  }
}));

// Get digest history
router.get('/history/:userId', asyncHandler(async (req: Request, res: Response) => {
  const { userId } = req.params;
  const { limit, offset } = req.query;
  
  console.log(`📚 Getting digest history for user ${userId}`);
  
  try {
    const digests = await storage.getEmailDigests(userId);
    
    // Apply pagination if provided
    let paginatedDigests = digests;
    if (limit) {
      const limitNum = parseInt(limit as string);
      const offsetNum = parseInt(offset as string) || 0;
      paginatedDigests = digests.slice(offsetNum, offsetNum + limitNum);
    }
    
    res.json(apiResponse({
      digests: paginatedDigests,
      total: digests.length,
      ...(limit && { 
        limit: parseInt(limit as string),
        offset: parseInt(offset as string) || 0
      })
    }));
    
  } catch (error: any) {
    console.error('Error getting digest history:', error);
    res.status(500).json(apiError(
      `Failed to get digest history: ${error.message}`,
      'HISTORY_RETRIEVAL_FAILED'
    ));
  }
}));

// Get digest by date
router.get('/date/:userId/:date', asyncHandler(async (req: Request, res: Response) => {
  const { userId, date } = req.params;
  
  console.log(`📅 Getting digest for user ${userId} on date ${date}`);
  
  try {
    // Check for thematic digest first
    const targetDate = new Date(date);
    const hasThematic = await storage.hasThematicDigestForDate(userId, targetDate);
    
    if (hasThematic) {
      const thematicDigests = await storage.getThematicDigests(userId);
      const thematicDigest = thematicDigests.find(digest => {
        const digestDateStr = new Date(digest.date).toISOString().split('T')[0];
        return digestDateStr === date;
      });
      
      if (thematicDigest) {
        const fullThematicDigest = await storage.getThematicDigest(userId, thematicDigest.id);
        if (fullThematicDigest) {
          return res.json(apiResponse({
            ...fullThematicDigest,
            type: 'thematic',
            date: fullThematicDigest.date instanceof Date ? 
              fullThematicDigest.date.toISOString() : fullThematicDigest.date,
            emailsProcessed: fullThematicDigest.totalSourceEmails,
            topicsIdentified: fullThematicDigest.sectionsCount
          }));
        }
      }
    }
    
    // Fall back to regular digest
    const digest = await storage.getDigestByDate(userId, date);
    
    if (!digest) {
      return res.status(404).json(apiError('No digest found for this date', 'NOT_FOUND'));
    }
    
    // Get emails for this digest
    const emails = await storage.getDigestEmails(digest.id);
    
    const response = {
      ...digest,
      type: 'regular',
      date: digest.date instanceof Date ? digest.date.toISOString() : digest.date,
      emails: emails.map(email => ({
        ...email,
        receivedAt: email.receivedAt instanceof Date ? 
          email.receivedAt.toISOString() : email.receivedAt
      }))
    };
    
    res.json(apiResponse(response));
    
  } catch (error: any) {
    console.error('Error getting digest by date:', error);
    res.status(500).json(apiError(
      `Failed to get digest for date: ${error.message}`,
      'DATE_DIGEST_FAILED'
    ));
  }
}));

// ==================== THEMATIC PROCESSING ====================

// Process emails into thematic digest
router.post('/thematic/process', asyncHandler(async (req: Request, res: Response) => {
  const { userId, emailDigestId, emails } = req.body;
  
  if (!userId || !emailDigestId || !emails || !Array.isArray(emails)) {
    return res.status(400).json(apiError(
      'userId, emailDigestId, and emails array are required', 
      'MISSING_FIELDS'
    ));
  }
  
  console.log(`🎯 Processing ${emails.length} emails into thematic digest for user ${userId}`);
  
  try {
    // Import thematic processor (would be moved to services)
    const { thematicProcessor } = await import('../services/thematic-processor.js');
    
    const thematicDigestId = await thematicProcessor.processEmailsIntoThemes(userId, emails);
    
    res.status(201).json(apiResponse({
      thematicDigestId,
      emailsProcessed: emails.length,
      userId
    }, 'Thematic digest created successfully'));
    
  } catch (error: any) {
    console.error('Error processing thematic digest:', error);
    res.status(500).json(apiError(
      `Failed to process thematic digest: ${error.message}`,
      'THEMATIC_PROCESSING_FAILED'
    ));
  }
}));

// ==================== ANALYTICS ====================

// Get digest statistics
router.get('/stats/:userId', asyncHandler(async (req: Request, res: Response) => {
  const { userId } = req.params;
  const { period } = req.query; // day, week, month, year
  
  console.log(`📊 Getting digest statistics for user ${userId}, period: ${period}`);
  
  try {
    const digests = await storage.getEmailDigests(userId);
    const thematicDigests = await storage.getThematicDigests(userId);
    
    // Calculate statistics
    const stats = {
      totalDigests: digests.length,
      totalThematicDigests: thematicDigests.length,
      totalEmailsProcessed: digests.reduce((sum, d) => sum + d.emailsProcessed, 0),
      totalTopicsIdentified: digests.reduce((sum, d) => sum + d.topicsIdentified, 0),
      averageEmailsPerDigest: digests.length > 0 ? 
        digests.reduce((sum, d) => sum + d.emailsProcessed, 0) / digests.length : 0,
      averageTopicsPerDigest: digests.length > 0 ?
        digests.reduce((sum, d) => sum + d.topicsIdentified, 0) / digests.length : 0,
      mostRecentDigest: digests.length > 0 ? digests[0].date : null,
      period: period || 'all-time'
    };
    
    res.json(apiResponse(stats));
    
  } catch (error: any) {
    console.error('Error getting digest statistics:', error);
    res.status(500).json(apiError(
      `Failed to get digest statistics: ${error.message}`,
      'STATS_RETRIEVAL_FAILED'
    ));
  }
}));

export { router as digestRoutes };